backend/migrations/versions/d05f9a5b215e_backfill_dataset_table_permissions.py

```diff
@@ -0,0 +1,131 @@
+    confidentiality = Column(String, nullable=False, default='C1')
+
+
+class ShareObjectItem(Base):
+    __tablename__ = 'share_object_item'
+    shareUri = Column(String, nullable=False)
+    shareItemUri = Column(
+        String, default=utils.uuid('shareitem'), nullable=False, primary_key=True
+    )
+    itemType = Column(String, nullable=False)
+    itemUri = Column(String, nullable=False)
+    itemName = Column(String, nullable=False)
+    permission = Column(String, nullable=True)
+    created = Column(DateTime, nullable=False, default=datetime.now)
+    updated = Column(DateTime, nullable=True, onupdate=datetime.now)
+    deleted = Column(DateTime, nullable=True)
+    owner = Column(String, nullable=False)
+    GlueDatabaseName = Column(String, nullable=True)
+    GlueTableName = Column(String, nullable=True)
+    S3AccessPointName = Column(String, nullable=True)
+    status = Column(String, nullable=False, default=ShareObjectStatus.Draft.value)
+    action = Column(String, nullable=True)
+
+
+def upgrade():
+    try:
+        bind = op.get_bind()
+        session = orm.Session(bind=bind)
+        print('Re-Initializing permissions...')
+        api.Permission.init_permissions(session)
+        print('Permissions re-initialized successfully')
+    except Exception as e:
+        print(f'Failed to init permissions due to: {e}')
+
+    try:
+        bind = op.get_bind()
+        session = orm.Session(bind=bind)
+        print('Back-filling dataset table permissions for owners/stewards...')
+        dataset_tables: [DatasetTable] = session.query(DatasetTable).filter(DatasetTable.deleted.is_(None)).all()
+        for table in dataset_tables:
+            dataset = api.Dataset.get_dataset_by_uri(session, table.datasetUri)
+            env = api.Environment.get_environment_by_uri(session, dataset.environmentUri)
+
+            groups = set([dataset.SamlAdminGroupName, env.SamlGroupName, dataset.stewards if dataset.stewards is not None else dataset.SamlAdminGroupName])
+            for group in groups:
+                api.ResourcePolicy.attach_resource_policy(
+                    session=session,
+                    resource_uri=table.tableUri,
+                    group=group,
+                    permissions=permissions.DATASET_TABLE_READ,
+                    resource_type=models.DatasetTable.__name__,
+                )
+        print('dataset table permissions updated successfully for owners/stewards')
+    except Exception as e:
+        print(f'Failed to backfill dataset table permissions for owners/stewards due to: {e}')
+
+    try:
+        bind = op.get_bind()
+        session = orm.Session(bind=bind)
+        print('Back-filling dataset table permissions for shared principals...')
+        share_table_items: [ShareObjectItem] = session.query(ShareObjectItem).filter(
+            (
+                and_(
+                    ShareObjectItem.status == ShareObjectStatus.Share_Succeeded.value,
```

Comment:
I like that the addition of permissions happens after the share has succeeded, in the implementation we grant when a table is approved, but I think it is a bit difficult to avoid doing it at a different stage.